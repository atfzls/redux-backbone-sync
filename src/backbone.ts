import Backbone from 'backbone';

// backbone start
export const Song = Backbone.Model.extend({});

export const Songs = Backbone.Collection.extend({
  model: Song,
});
