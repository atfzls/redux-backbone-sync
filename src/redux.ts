import produce from "immer";
import { createStore } from "redux";

export const store = createStore((state: any, action: any) => {
  switch (action.type) {
    case "INCREASE_LISTENER": {
      const id = action.payload;
      return produce(state, (draft: any) => {
        draft.songs.find((song: any) => song.id === id).listeners += 1;
      });
    }
    case "UPDATE_SONG_TITLE": {
      const id = action.payload.id;
      return produce(state, (draft: any) => {
        draft.songs.find((song: any) => song.id === id).title =
          action.payload.title;
      });
    }
    case "EVAL": {
      return action.payload(state);
    }
    default: {
      return state;
    }
  }
}, {});
