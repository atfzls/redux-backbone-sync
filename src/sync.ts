import * as _ from "lodash";
import { Store } from "redux";
import produce from "immer";

export function sync(
  store: Store<any, any>,
  model: Backbone.Collection,
  slicePath: string
) {
  const updateStore = () =>
    store.dispatch({
      type: "EVAL",
      payload: (state: any) => {
        return produce(state, (draft: any) => {
          _.set(draft, slicePath, model.toJSON());
        });
      }
    });

  updateStore(); // for hydrating store with initial state
  const callback = () => {
    updateStore();
  };
  const unsubscribeModel = () => {
    model.off("change", callback);
  };
  model.on("change", () => {
    updateStore();
  });

  const subscribeToSlice = (key: string) => {
    let oldSlice: any;

    return store.subscribe(() => {
      const slice = _.get(store.getState(), key);

      if (oldSlice !== slice) {
        model.reset();
        model.set(slice);
        oldSlice = slice;
      }
    });
  };

  const unsubscribeStore = subscribeToSlice(slicePath);

  return [unsubscribeModel, unsubscribeStore];
}
