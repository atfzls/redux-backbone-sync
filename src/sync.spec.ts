import { store } from "./redux";
import { Songs, Song } from "./backbone";
import { sync } from "./sync";

describe("sync backbone model and redux store", () => {
  let songs: any;
  let disposables: Function[] = [];

  beforeEach(() => {
    songs = new Songs([
      new Song({ title: "Blue in Green", listeners: 0, id: 1 }),
      new Song({ title: "So What", listeners: 0, id: 2 }),
      new Song({ title: "All Blues", listeners: 0, id: 3 })
    ]);
  
    disposables = sync(store, songs, 'songs');
  });

  afterEach(() => {
    disposables.forEach(disposable => disposable());
  });

  describe('update redux after updating model', () => {
    it("updating title of song ", () => {
      songs.at(0).set('title', 'New Title');
      expect(store.getState()).toMatchSnapshot();
    });
  
    it("updating listeners of song ", () => {
      songs.at(0).set('listeners', songs.at(0).get('listeners') + 1);
      expect(store.getState()).toMatchSnapshot();
    });
  
    it("updating multiple listeners of song ", () => {
      songs.at(0).set('listeners', songs.at(0).get('listeners') + 1);
      songs.at(0).set('listeners', songs.at(0).get('listeners') + 1);
      songs.at(0).set('listeners', songs.at(0).get('listeners') + 1);
  
      songs.at(1).set('listeners', songs.at(1).get('listeners') + 1);
      songs.at(2).set('listeners', songs.at(2).get('listeners') + 1);
  
      expect(store.getState()).toMatchSnapshot();
    });
  });

  describe('update backbone model after updating redux state', () => {
    it('updating title of song', () => {
      store.dispatch({
        type: 'UPDATE_SONG_TITLE',
        payload: {
          id: 2,
          title: 'New Title for id 2'
        }
      });

      expect(songs.toJSON()).toMatchSnapshot();
    });

    it('updating listeners', () => {
      store.dispatch({
        type: 'INCREASE_LISTENER',
        payload: 1
      });

      expect(songs.toJSON()).toMatchSnapshot();
    });

    it('updating multiple listeners of song', () => {
      store.dispatch({
        type: 'INCREASE_LISTENER',
        payload: 1
      });
      store.dispatch({
        type: 'INCREASE_LISTENER',
        payload: 1
      });
      store.dispatch({
        type: 'INCREASE_LISTENER',
        payload: 1
      });
      store.dispatch({
        type: 'INCREASE_LISTENER',
        payload: 2
      });
      store.dispatch({
        type: 'INCREASE_LISTENER',
        payload: 3
      });

      expect(songs.toJSON()).toMatchSnapshot();
    });
  });
});
