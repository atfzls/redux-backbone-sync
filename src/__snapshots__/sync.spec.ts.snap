// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sync backbone model and redux store update backbone model after updating redux state updating listeners 1`] = `
Array [
  Object {
    "id": 1,
    "listeners": 1,
    "title": "Blue in Green",
  },
  Object {
    "id": 2,
    "listeners": 0,
    "title": "So What",
  },
  Object {
    "id": 3,
    "listeners": 0,
    "title": "All Blues",
  },
]
`;

exports[`sync backbone model and redux store update backbone model after updating redux state updating multiple listeners of song 1`] = `
Array [
  Object {
    "id": 1,
    "listeners": 3,
    "title": "Blue in Green",
  },
  Object {
    "id": 2,
    "listeners": 1,
    "title": "So What",
  },
  Object {
    "id": 3,
    "listeners": 1,
    "title": "All Blues",
  },
]
`;

exports[`sync backbone model and redux store update backbone model after updating redux state updating title of song 1`] = `
Array [
  Object {
    "id": 1,
    "listeners": 0,
    "title": "Blue in Green",
  },
  Object {
    "id": 2,
    "listeners": 0,
    "title": "New Title for id 2",
  },
  Object {
    "id": 3,
    "listeners": 0,
    "title": "All Blues",
  },
]
`;

exports[`sync backbone model and redux store update redux after updating model updating listeners of song  1`] = `
Object {
  "songs": Array [
    Object {
      "id": 1,
      "listeners": 1,
      "title": "Blue in Green",
    },
    Object {
      "id": 2,
      "listeners": 0,
      "title": "So What",
    },
    Object {
      "id": 3,
      "listeners": 0,
      "title": "All Blues",
    },
  ],
}
`;

exports[`sync backbone model and redux store update redux after updating model updating multiple listeners of song  1`] = `
Object {
  "songs": Array [
    Object {
      "id": 1,
      "listeners": 3,
      "title": "Blue in Green",
    },
    Object {
      "id": 2,
      "listeners": 1,
      "title": "So What",
    },
    Object {
      "id": 3,
      "listeners": 1,
      "title": "All Blues",
    },
  ],
}
`;

exports[`sync backbone model and redux store update redux after updating model updating title of song  1`] = `
Object {
  "songs": Array [
    Object {
      "id": 1,
      "listeners": 0,
      "title": "New Title",
    },
    Object {
      "id": 2,
      "listeners": 0,
      "title": "So What",
    },
    Object {
      "id": 3,
      "listeners": 0,
      "title": "All Blues",
    },
  ],
}
`;
